packages/langium/src/parser/token-builder.ts

```diff
@@ -29,6 +29,9 @@ export class DefaultTokenBuilder implements TokenBuilder {
         const reachableRules = stream(getAllReachableRules(grammar, false));
         const terminalTokens: TokenType[] = this.buildTerminalTokens(reachableRules);
         const tokens: TokenType[] = this.buildKeywordTokens(reachableRules, terminalTokens, options);
+        if(reachableRules.some(r => streamAst(r.definition).some(isEndOfFile))) {
+            tokens.push(EOF);
+        }
```

Comment:
Suggestion: The `EOF` token type should be pushed to the end of the lexer definition. Otherwise the lexer will always try to match it first. Given that it's a fairly unlikely token to encounter, we should move it to the end of the lexer definition.